File: 3670.maximum-product-of-two-integers-with-no-common-bits.java
#
# @lc app=leetcode id=3670 lang=java
#
# [3670] Maximum Product of Two Integers With No Common Bits
#
# @lc code=start
class Solution {
    private static class TrieNode {
        TrieNode[] children = new TrieNode[2];
    }

    public long maxProduct(int[] nums) {
        TrieNode root = new TrieNode();
        for (int num : nums) {
            insert(root, num);
        }
        long ans = 0;
        for (int num : nums) {
            int cand = getMaxCompatible(root, num);
            ans = Math.max(ans, (long) num * cand);
        }
        return ans;
    }

    private void insert(TrieNode root, int num) {
        TrieNode curr = root;
        for (int i = 19; i >= 0; i--) {
            int bit = (num >> i) & 1;
            if (curr.children[bit] == null) {
                curr.children[bit] = new TrieNode();
            }
            curr = curr.children[bit];
        }
    }

    private int getMaxCompatible(TrieNode root, int x) {
        TrieNode curr = root;
        int y = 0;
        for (int i = 19; i >= 0; i--) {
            int bitX = (x >> i) & 1;
            int bitY = 0;
            TrieNode next = null;
            if (bitX == 1) {
                next = curr.children[0];
            } else {
                if (curr.children[1] != null) {
                    bitY = 1;
                    next = curr.children[1];
                } else if (curr.children[0] != null) {
                    next = curr.children[0];
                }
            }
            if (next == null) {
                return 0;
            }
            y |= (bitY << i);
            curr = next;
        }
        return y;
    }
}
# @lc code=end

File: 3670.maximum-product-of-two-integers-with-no-common-bits.go
#
# @lc app=leetcode id=3670 lang=golang
#
# [3670] Maximum Product of Two Integers With No Common Bits
#

# @lc code=start
func maxProduct(nums []int) int64 {
	type Node struct {
		children [2]*Node
	}

	root := &Node{}

	insert := func(nd *Node, val int) {
		node := nd
		for i := 19; i >= 0; i-- {
			bit := (val >> i) & 1
			if node.children[bit] == nil {
				node.children[bit] = &Node{}
			}
			node = node.children[bit]
		}
	}

	getMax := func(nd *Node, val int) int {
		node := nd
		res := 0
		for i := 19; i >= 0; i-- {
			bit := (val >> i) & 1
			if bit == 1 {
				if node.children[0] == nil {
					return 0
				}
				node = node.children[0]
			} else {
				if node.children[1] != nil {
					res |= 1 << i
					node = node.children[1]
				} else if node.children[0] != nil {
					node = node.children[0]
				} else {
					return 0
				}
			}
		}
		return res
	}

	ans := int64(0)
	for _, num := range nums {
		partner := getMax(root, num)
		curr := int64(num) * int64(partner)
		if curr > ans {
			ans = curr
		}
		insert(root, num)
	}
	return ans
}
# @lc code=end